include/seqan3/search/dream_index/detail/binning_directory_base.hpp

```diff
@@ -0,0 +1,209 @@
+ * \brief Provides the seqan3::detail::binning_directory_base.
+ */
+
+#pragma once
+
+#include <sdsl/bit_vectors.hpp>
+
+#include <seqan3/core/concept/cereal.hpp>
+#include <seqan3/std/algorithm>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\brief The IBF binning directory.
+ * \tparam data_type The underlying data type.
+ * \implements seqan3::BinningDirectory
+ */
+template <typename data_type>
+class binning_directory_base
+{
+protected:
+    //!\brief The number of bins.
+    size_t bins{};
+    //!\brief The number of 64 bit integers used to represent the numerical value `bins`.
+    size_t bin_width{};
+    //!\brief How big is bins in multiples of 64.
+    size_t block_size{};
+    //!\brief How many blocks fit in the bitvector.
+    size_t block_count{};
+    //!\brief The number of hash functions.
+    size_t num_hash{};
+    //!\brief The bitvector.
+    data_type data{};
+    //!\brief Precalculated values for hashing.
+    std::vector<size_t> pre_hash{};
+    //!\brief Shift value for hashing.
+    static constexpr size_t shift{27};
+    //!\brief Seed for hashing.
+    static constexpr size_t seed{0x90b45d39fb6da1fa};
+
+    /*!\brief Perturbs a value and fits it into the vector.
+     * \param h The value to process.
+     */
+    inline constexpr void hash_and_fit(size_t & h) const
+    {
+        h ^= h >> shift; // Basically Fibonacci hashing
+        h %= block_count;
+        h *= block_size;
+    }
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    binning_directory_base() = default;                                           //!< Defaulted.
+    binning_directory_base(binning_directory_base const &) = default;             //!< Defaulted.
+    binning_directory_base & operator=(binning_directory_base const &) = default; //!< Defaulted.
+    binning_directory_base(binning_directory_base &&) = default;                  //!< Defaulted.
+    binning_directory_base & operator=(binning_directory_base &&) = default;      //!< Defaulted.
+    ~binning_directory_base() = default;                                          //!< Defaulted.
+
+    /*!\brief Returns the number of bins in the binning directory.
+     * \returns The number of bins.
+     */
+    inline size_t get_bins() const
+    {
+        return bins;
+    }
+
+    /*!\brief Returns the number of hash functions used in the binning directory.
+     * \returns The number of hash functions.
+     */
+    inline size_t get_num_hash() const
+    {
+        return num_hash;
+    }
+
+    /*!\brief Determines set membership of a given value.
+     * \param[in,out] result The sdsl::bit_vector to store the result into.
+     * \param[in]     h      The raw hash value to process.
+     * \returns A sdsl::bit_vector of size bins where each position indicates the bin membership of the hash value.
+     */
+    [[nodiscard]] sdsl::bit_vector get(size_t const h, sdsl::bit_vector && result) const
```

Comment:
can't you default the result and save a function overload?
```suggestion
    [[nodiscard]] sdsl::bit_vector get(size_t const h, sdsl::bit_vector && result = sdsl::bit_vector(bins)) const
```
If not the overload below should at least use this overload instead of the code duplication
```cpp
    [[nodiscard]] sdsl::bit_vector get(size_t const h) const
    {
        sdsl::bit_vector result(bins);
        get(h, result);
    }
```

Reply:
I offer
```cpp
    [[nodiscard]] sdsl::bit_vector get(size_t const h) const
    {
        return get(h, sdsl::bit_vector(bins));
    }
```

I can't default it since the default depends on a non-static member:
https://godbolt.org/z/zanlhF

Reply:
:+1: